parquet/src/main/scala/magnolify/parquet/Predicate.scala

```diff
@@ -0,0 +1,84 @@
+package magnolify.parquet
+
+import magnolify.shared.CaseMapper
+import org.apache.parquet.filter2.predicate.Operators.Column
+import org.apache.parquet.filter2.predicate.{
+  FilterApi,
+  FilterPredicate,
+  Statistics,
+  UserDefinedPredicate
+}
+import org.apache.parquet.io.api.{Binary, PrimitiveConverter}
+import org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName
+
+object Predicate {
+
+  /**
+   * Constructs a Parquet [[FilterPredicate]] based on a single primitive column and a user-defined
+   * function. Predicates can be logically combined using [[FilterApi.and]] and [[FilterApi.or]].
+   *
+   * Note that Optional fields should use a predicate for their lifted type: that is, create a
+   * predicate for an optional String field using [[Predicate.onField[String]], even if its
+   * projected case class defines the field as `Option[String]`. When applying predicates, Parquet
+   * will filter out optional fields with a value of `null` before applying the predicate filter.
+   *
+   * @param fieldName
+   *   the name of the primitive column. Nested fields are separated by '.'s
+   * @param filterFn
+   *   the UDF representing the desired filter
+   * @return
+   *   a FilterPredicate for use with typed Parquet
+   */
+  def onField[ScalaFieldT](
+    fieldName: String
+  )(filterFn: ScalaFieldT => Boolean)(implicit
+    pf: ParquetField.Primitive[ScalaFieldT]
+  ): FilterPredicate = {
+    val fieldType = pf.schema(CaseMapper.identity).asPrimitiveType().getPrimitiveTypeName
+
+    val column = fieldType match {
+      case PrimitiveTypeName.INT32                           => FilterApi.intColumn(fieldName)
+      case PrimitiveTypeName.INT64 | PrimitiveTypeName.INT96 => FilterApi.longColumn(fieldName)
+      case PrimitiveTypeName.BINARY                          => FilterApi.binaryColumn(fieldName)
+      case PrimitiveTypeName.FLOAT                           => FilterApi.floatColumn(fieldName)
+      case PrimitiveTypeName.DOUBLE                          => FilterApi.doubleColumn(fieldName)
+      case PrimitiveTypeName.BOOLEAN                         => FilterApi.booleanColumn(fieldName)
+      case _ => throw new UnsupportedOperationException(s"Unsupported column type $fieldType")
+    }
+
+    def wrap[T](addFn: (PrimitiveConverter, T) => Unit): T => ScalaFieldT = {
+      val converter = pf.newConverter
+      value => {
+        addFn(converter.asPrimitiveConverter(), value)
+        converter.get
+      }
+    }
+
+    type PC = PrimitiveConverter
+    val toScalaT = (fieldType match {
```

Comment:
I'm torn about this converter implementation, because deferring to the `ParquetField#newConverter` means we're not duplicating any work here, but for some field types, like Integer, it seems wasteful to go through the process of buffering the bytes and creating a `PrimitiveConverter` instance when it can just be cast to the Scala type.

Reply:
You could do this with `ScalaFieldT : ClassTag` and a check that the types are compatible for these few cases?

Reply:
yeah that would work! I just did some benchmarking on 300gb of a Parquet dataset and see literally no performance difference between these two predicates:

```scala
val magnolifyPred = Predicate.onField[Long]("field")(_ % 2 == 0))

val manualPred = FilterApi.userDefined[java.lang.Long, UserDefinedPredicate[java.lang.Long] with Serializable](
        FilterApi.longColumn("field"),
        new UserDefinedPredicate[java.lang.Long] with Serializable {
        override def keep(value: lang.Long): Boolean = value % 2 == 0

        override def canDrop(statistics: Statistics[lang.Long]): Boolean = false

        override def inverseCanDrop(statistics: Statistics[lang.Long]): Boolean = false
      }))
```

So I guess we can leave it as is, for now anyway. I think the Converters recycle their buffers so it's not as wasteful as I thought.